### ISSUE TITLE
[EPIC][SECURITY]: ML-Based Behavioral Fingerprinting for Issue #257

### ISSUE THREAD

**Issue by crivetimihai** [IBM/mcp-context-forge#2603]
# Enhancement Proposal for Issue #257

## Original Issue

**Title**: `[FEATURE][SECURITY]: Gateway-level rate limiting, DDoS protection, and abuse detection`

**Current Scope**: Comprehensive protection framework with:
- Token bucket rate limiting
- DDoS detection (volumetric, geographic)
- Behavioral abuse detection (resource exhaustion, credential stuffing)
- Redis-backed distributed counters
- Prometheus metrics

---

### Summary

Extend #257's abuse detection with **machine learning-based behavioral analysis** that learns normal usage patterns per client/tenant and detects anomalies that rule-based systems miss. This adds adaptive threat detection that evolves with attack patterns.

### Why Extend #257?

The current #257 covers:
- **Rule-based detection**: X failures in Y minutes = brute force
- **Threshold-based DDoS**: 10x traffic spike = attack
- **Pattern matching**: Known abuse signatures

This enhancement adds:
- **Learned baselines**: What's normal for this specific client?
- **Anomaly scoring**: How unusual is this request pattern?
- **Adaptive thresholds**: Auto-adjust based on historical behavior
- **Unknown attack detection**: Catch novel attack patterns

---

## Additional User Story

<details>
<summary><strong>US-4: SOC Analyst - ML-Based Anomaly Detection</strong></summary>

**As a** SOC Analyst
**I want** the gateway to detect unusual behavior patterns
**So that** I can catch sophisticated attacks that evade rule-based detection

**Acceptance Criteria:**

```gherkin
Given the ML model has learned baseline behavior for client "api_client_123":
  - Normal: 100-200 requests/hour during business hours
  - Normal: Access to tools A, B, C
  - Normal: Request sizes 1-10KB

When client behavior deviates significantly:
  - 500 requests in 10 minutes at 3 AM
  - Access to admin_* tools never used before
  - Request payloads averaging 500KB

Then:
  - Anomaly score calculated: 0.87 (high)
  - Alert triggered with behavioral breakdown
  - Request allowed but flagged for review (soft block)
  - Enhanced logging enabled for this session

Given a new attack pattern emerges:
When multiple clients show similar anomalous behavior:
Then:
  - Pattern is clustered as potential coordinated attack
  - SOC alerted to review cluster
  - Temporary rule can be created from pattern
```

</details>

---

### ML Detection Flow

```mermaid
flowchart TD
    A[Incoming Request] --> B[Feature Extraction]
    B --> C[Load Client Baseline]
    C --> D{Baseline Exists?}

    D -->|No| E[Use Global Baseline]
    D -->|Yes| F[Calculate Deviation]

    E --> G[Anomaly Scoring]
    F --> G

    G --> H{Score > Threshold?}
    H -->|No| I[Allow + Update Baseline]
    H -->|Yes| J{Score > Hard Block?}

    J -->|No| K[Allow + Alert + Enhanced Logging]
    J -->|Yes| L[Block + Alert + Incident]

    I --> M[Request Processing]
    K --> M
    L --> N[Return 429/403]

    subgraph "Baseline Learning"
        O[Request Features] --> P[Feature Store]
        P --> Q[Baseline Model]
        Q --> R[Client Profiles]
    end

    subgraph "Features Extracted"
        F1[Request Rate]
        F2[Time of Day]
        F3[Endpoint Distribution]
        F4[Payload Size]
        F5[User Agent Patterns]
        F6[Geographic Location]
        F7[Tool Usage Patterns]
    end
```

### Feature Vector

```python
@dataclass
class BehaviorFeatures:
    """Features extracted from request for ML analysis."""

    # Rate features
    requests_per_minute: float
    requests_per_hour: float
    burst_ratio: float  # max/avg request rate

    # Temporal features
    hour_of_day: int
    day_of_week: int
    is_business_hours: bool
    time_since_last_request: float

    # Access pattern features
    endpoint_entropy: float  # diversity of endpoints
    unique_endpoints_1h: int
    new_endpoint_ratio: float  # % never-seen endpoints

    # Payload features
    avg_payload_size: float
    payload_size_variance: float

    # Tool usage features
    tool_diversity: float
    admin_tool_ratio: float
    new_tool_ratio: float

    # Network features
    ip_reputation_score: float
    geo_anomaly_score: float  # vs historical locations
    user_agent_consistency: float
```

---

### Phase 5: ML Behavioral Analysis (NEW)

- [ ] **Feature extraction**
  - [ ] Create `BehaviorFeatures` dataclass
  - [ ] Implement feature extractors (no external API calls on request path)
  - [ ] Local GeoIP database for geo features (MaxMind)
  - [ ] Cached IP reputation (async refresh)
  - [ ] Redis-based feature store per client
  - [ ] Define retention policy for behavioral data

- [ ] **Baseline learning**
  - [ ] Start with EMA + z-score (simple, fast)
  - [ ] Global baseline for new clients (until 500+ requests or 7 days)
  - [ ] Configurable learning rate
  - [ ] Confidence score based on sample size

- [ ] **Anomaly detection**
  - [ ] Phase 1: Statistical (z-score, EMA deviation)
  - [ ] Phase 2: Isolation Forest (offline trained, real-time scoring)
  - [ ] Anomaly scoring (0-1 scale) with confidence
  - [ ] Configurable thresholds per severity level

- [ ] **Performance optimizations**
  - [ ] Async scoring (non-blocking)
  - [ ] Decision caching (30-60s TTL per client)
  - [ ] Strict latency budget (skip if exceeded)
  - [ ] Offline batch training for models

- [ ] **Alert integration**
  - [ ] Anomaly alerts with feature breakdown
  - [ ] Cluster detection for coordinated attacks
  - [ ] Integration with existing alerting

### Phase 6: Feedback Loop (NEW)

- [ ] **SOC feedback integration**
  - [ ] Mark alerts as true/false positive
  - [ ] Retrain model with feedback
  - [ ] Threshold tuning recommendations

- [ ] **Rule generation**
  - [ ] Generate rule proposals from patterns
  - [ ] Admin approval workflow
  - [ ] Automatic rule deployment

---

## Configuration Extension

```yaml
# Addition to #257 configuration
abuse_detection:
  # Existing rule-based config...

  # NEW: ML behavioral analysis
  ml_behavioral:
    enabled: true

    # Baseline learning
    learning:
      enabled: true
      learning_rate: 0.1
      min_samples: 100  # requests before client-specific baseline
      global_baseline_percentile: 50

    # Anomaly detection
    detection:
      algorithm: isolation_forest
      contamination: 0.01  # expected anomaly rate
      soft_block_threshold: 0.7
      hard_block_threshold: 0.9

    # Features to analyze
    features:
      - request_rate
      - temporal_patterns
      - endpoint_distribution
      - payload_characteristics
      - tool_usage_patterns

    # Alerting
    alerting:
      anomaly_threshold: 0.6
      cluster_detection: true
      cluster_min_size: 3

    # Feedback
    feedback:
      enabled: true
      auto_retrain_interval_hours: 24
```

---

## Success Criteria (Additional)

- [ ] Anomaly detection catches patterns rule-based misses
- [ ] < 5% false positive rate on anomaly alerts
- [ ] Client baselines established within 100 requests
- [ ] Cluster detection identifies coordinated attacks
- [ ] Feedback loop improves detection over time

---

## Definition of Done (Additional)

- [ ] Feature extraction implemented
- [ ] Baseline learning (per-client + global)
- [ ] Anomaly scoring with Isolation Forest
- [ ] Alert integration with feature breakdown
- [ ] Feedback loop for model improvement
- [ ] Admin UI for anomaly review
- [ ] Unit tests for feature extraction
- [ ] Integration tests for detection accuracy

---

## Notes

**Computational Overhead**: ML scoring adds ~5-10ms per request. Mitigations:
- Async scoring (score after response, cache decision for next request)
- Strict latency budget (skip ML if budget exceeded)
- Cache recent decisions per client (TTL 30-60s)
- Offline batch scoring for model training

**External Lookup Latency**: Feature extraction for geo/IP reputation requires:
- Local GeoIP database (MaxMind, IP2Location) - no external calls
- Cached IP reputation (refresh async, not per-request)
- Or: Skip these features for real-time scoring, use in offline analysis

**Model Simplicity**: Start with lightweight statistical methods:
- Exponential Moving Average (EMA) for baseline
- Z-score for anomaly detection
- Graduate to Isolation Forest only if needed
- Deep learning is overkill for this use case

**Baseline Requirements**: "100 requests" may be too few for stable baseline:
- Consider 500+ requests or 7 days of activity
- Use global baseline for new clients until sufficient data
- Flag confidence level in anomaly scores

**Tenant Isolation**: Cross-tenant clustering for coordinated attack detection:
- Must anonymize/pseudonymize before aggregation
- Aggregate by behavior pattern, not tenant identity
- Consider privacy implications and data retention

**Per-Tenant Model Cost**: Isolation Forest per-tenant is expensive:
- Train offline in batch jobs, not real-time
- Share model structure, per-tenant baselines only
- Consider global model with tenant-specific thresholds

**Privacy**: Behavioral features should not include PII:
- Client ID is pseudonymous
- Request content is NOT analyzed
- Define retention policy for behavioral data
- Clear pseudonymization for stored features

---

## References

- [Isolation Forest for Anomaly Detection](https://scikit-learn.org/stable/modules/generated/sklearn.ensemble.IsolationForest.html)
- [Behavioral Analysis in Web Application Firewalls](https://www.cloudflare.com/learning/ddos/glossary/web-application-firewall-waf/)
- [Netflix's Anomaly Detection at Scale](https://netflixtechblog.com/rad-outlier-detection-on-big-data-d6b0494371cc)
